content: How Mercado Libre optimized for Web Vitals (TBT, FID) (#3814)
GoogleChrome/web.dev: src/site/content/en/blog/how-mercadolibre-optimized-web-vitals/index.md

---
layout: post
title: How Mercado Libre optimized for Web Vitals (TBT/FID)
subhead: Optimizing interactivity of Mercado Libre's product details pages.
authors:
  - demianrenzulli
  - aranhacarlos
  - joanbaca
date: 2020-09-01

> hero: hero.jpg

Reviewer: The hero image looks cutoff on web.dev/blog in an awkward way:

![image](https://user-images.githubusercontent.com/4713486/92967924-f25bd900-f42e-11ea-9c9b-96a1f16bcf83.png)

Do you want to create a thumbnail that focuses on the Mercado Libre logo? https://web.dev/handbook/markup-media/#thumbnail

description: Summary of the work done by Mercado Libre's frontend architecture team to optimize FID, using TBT as a proxy metric.
tags:
  - blog
  - performance

>   - web-vitals
> ---

Reviewer: ```suggestion
  - web-vitals
  - case-study
  - lighthouse
  - chrome-ux-report
  - javascript
---
```

Mercado Libre is the largest e-commerce and payments ecosystem in Latin America. It is present in 18 countries and is a market leader in Brazil, Mexico, and Argentina (based on unique visitors and pageviews).

> Web performance has been a focus for the company for a long time. They recently formed a team to monitor performance and apply optimizations across different parts of the site.

Reviewer: ```suggestion
Web performance has been a focus for the company for a long time, but they recently formed a team to monitor performance and apply optimizations across different parts of the site.
```

I noticed this in the draft but didn't know how to address it. But the answer is obvious now. We need to make the ideas in these two sentences flow into each other more. My suggested rephrasing does that. They created a dedicated performance team, but they were already focused on performance for a long time.

> This article summarizes the work done by [Guille Paz](https://twitter.com/pazguille), [Pablo Carminatti](https://www.linkedin.com/in/pcarminatti/), and [Oleh Burkhay](https://twitter.com/oburkhay) from Mercado Libre's frontend architecture team to optimize one of the Core Web Vitals: [First Input Delay (FID)](https://web.dev/fid/) and it's lab proxy, [Total Blocking Time (TBT)](https://web.dev/tbt/).

Reviewer: ```suggestion
This article summarizes the work done by [Guille Paz](https://twitter.com/pazguille), [Pablo Carminatti](https://www.linkedin.com/in/pcarminatti/), and [Oleh Burkhay](https://twitter.com/oburkhay) from Mercado Libre's frontend architecture team to optimize one of the Core Web Vitals: [First Input Delay (FID)](https://web.dev/fid/) and its lab proxy, [Total Blocking Time (TBT)](https://web.dev/tbt/).
```
it's = it is (which is not what we mean here)

> 

Reviewer: Can you add a [Stats component](https://web.dev/handbook/web-dev-components/#stats) here that summarizes the main successes?

Author: I've added two stats, please let me know if they look good.

Reviewer: LGTM

To provide a good user experience, sites should strive to have a First Input Delay of less than 100 milliseconds:

<picture>
  <source srcset="../vitals/fid_8x2.svg" media="(min-width: 640px)">
  <img class="w-screenshot w-screenshot--filled"
      src="../vitals/fid_4x3.svg"
      alt="Good fid values are 2.5 seconds, poor values are greater than 4.0
            seconds and anything in between needs improvement">
</picture>

While Mercado Libre's site was performing well in most sections, they found in the [Chrome User Experience Report](https://developers.google.com/web/tools/chrome-user-experience-report) that product detail pages had a poor FID. Based on that information, they decided to focus their efforts on improving the interactivity for product pages in the site.

<figure class="w-figure">
  <img src="meli-pdp.png" 
       alt="The mobile and desktop versions of a Mercado Libre product detail page.">
   <figcaption class="w-figcaption">
      The mobile and desktop versions of a Mercado Libre product detail page.
  </figcaption>     

> </figure>

Reviewer: Can you split this up into two images? The [Column compontent](https://web.dev/handbook/web-dev-components/#columns) should help you keep them side-by-side.

And make sure to add `class="w-screenshot"` to each image please

Author: Thanks @kaycebasques! I tried the following:

- Breaking the main image into two independent images (one for mobile the other for desktop).
- Using the column component.

In the original version, the mobile screenshot takes around 1/3 of the width. When using the component it assigned half of the width to each image, making the mobile one look much bigger and the desktop one very small.

I played around adding a width attribute to the img element, but it doesn't work either. My options here would be:

- Leaving the image as is.
- Picking only one of the two form factors and cut the other. I think the mobile one would be representative enough.

Let me know what you think.

Author: In the original image, mobile occupies 1/3 of the image and desktop 2/3. When breaking into two images, the column component assigns 50% to each, which makes the mobile look much bigger and the desktop much smaller.

I managed to make both look better by adding width="141" to the img element of meli-pdp-mobile.png, but it looks like a hack (perhaps you know a better way of doing this).

In any case, I still like more how it looked like in the original one.

Let me know if you have any ideas for this.

Reviewer: The columns you've provided look fine. I mainly wanted to split it up in order to add `class="w-screenshot"` to each

> These pages allow the user to perform complex interactions, so the goal was optimization interactivity, without removing valuable functionality.

Reviewer: ```suggestion
These pages allow the user to perform complex interactions, so the goal was interactivity optimization, without interfering with valuable functionality.
```

Here's the general approach they took:

- Use [WebPageTest](https://www.webpagetest.org/) to determine exactly which scripts were keeping the main thread busy on a real device.
- Use [Lighthouse](https://developers.google.com/web/tools/lighthouse) to determine the impact of the changes in [Max Potential First Input Delay (Max Potential FID)](https://web.dev/lighthouse-max-potential-fid/).

{% Aside %}
During this project Mercado Libre used [Max Potential FID](https://web.dev/lighthouse-max-potential-fid/) in Lighthouse because that was the tool's main metric for measuring interactivity at that time. Lighthouse now recommends using [Total Blocking Time](https://web.dev/tbt/) instead.
{% endAside %}

## Use WebPageTest to visualize long tasks

WebPageTest (WPT) is a web performance tool that allows you to run tests on real devices in different locations around the world.

Mercado Libre used WPT to reproduce the experience of their users by choosing a device type and location similar to real users. Specifically, they chose a **Moto 4G device** and **Dulles, Virginia**, because they wanted to approximate the experience of Mercado Libre users in Mexico.

> By observing the main thread view of WPT, Mercado Libre found that there were several consecutive long tasks blocking the man thread for 2 seconds:

Reviewer: ```suggestion
By observing the main thread view of WPT, Mercado Libre found that there were several consecutive long tasks blocking the main thread for 2 seconds:
```

<figure class="w-figure">

>   <img src="main-thread-unoptimized.png" 

Reviewer: ```suggestion
  <img src="main-thread-unoptimized.png" class="w-screenshot"
```

       alt="Main thread view of Mercado Libre's product detail pages.">
   <figcaption class="w-figcaption">
      Main thread view of Mercado Libre's product detail pages.
  </figcaption>     
</figure>

> Analyzing the corresponding waterfall they found that:
> 
> A considerable part of those two seconds came from their analytics module.
> The main bundle size of the application was large (950KB) and took a long time to parse, compile, and execute.

Reviewer: ```suggestion
Analyzing the corresponding waterfall they found that
a considerable part of those two seconds came from their analytics module.
The main bundle size of the application was large (950KB) and took a long time to parse, compile, and execute.
```

The colon (`they found that:`) makes it seem like a list will follow

<figure class="w-figure">

>   <img src="waterfall-unoptimized.png" 

Reviewer: ```suggestion
  <img src="waterfall-unoptimized.png" class="w-screenshot"
```

Mercado Libre used [webpack-bundle-analyzer](https://github.com/webpack-contrib/webpack-bundle-analyzer) to detect opportunities for optimization:

- Initially they were requiring the full [Lodash module](https://lodash.com/). This was replaced with a [per-method require](https://lodash.com/per-method-packages) to load only a subset of Lodash instead of the whole library, and used in conjunction with [lodash-webpack-plugin](https://github.com/lodash/lodash-webpack-plugin) to shrink Lodash even further.

They also applied the following [Babel](https://babeljs.io/) optimizations:

- Using [@babel/plugin-transform-runtime](https://babeljs.io/docs/en/babel-plugin-transform-runtime) to reuse Babel's helpers throughout the code, and reduce the size of the bundle considerably.
- Using [babel-plugin-search-and-replace](https://github.com/jean-smaug/babel-plugin-search-and-replace#readme) to replace tokens at build time, in order to remove a large configuration file inside the main bundle.
- Adding [babel-plugin-transform-react-remove-prop-types](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types#readme) to save some extra bytes by removing the prop types.

> As a result of these optimizations, the bundle size was reduced **by approximately 16%**.

Reviewer: Is this in addition to the 2% reduction that you mentioned earlier (so 18% total)? We should clarify this.

Author: Hi @kaycebasques! I know this part is confusing... In summary:

- The first optimization obtained a 2% reduction in JS for the whole site for **the internal tracking module**.
- The second optimization, obtained a 16% reduction in JS For **the main bundle loaded by PDP pages**.

I don't think we should consider that these optimizations should be added as 2% + 16% = 18%, as they have been applied to different assets, but, form your comments, the way it's expressed right now is not clear.

Maybe you have a suggestion on how to structure this part better?

Thanks a lot!

Author: These are two independent optimizations, and shouldn't be added as they were applied to the same asset:

- The 2% reduction is for the tracking module (this impacted the hole site as many sections use it).
- The 16% reduction was for the main bundle used. by PDP pages.

Please, let me know if you think there's a more clear way to explain this.

Reviewer: OK, makes sense. Your current phrasing works.

## Measure impact

The changes lowered Mercado Libre's consecutive long tasks **from two seconds to one second**:

<figure class="w-figure">

>   <img src="main-thread-iteration-1.png" 

Reviewer: ```suggestion
  <img src="main-thread-iteration-1.png" class="w-screenshot"
```

       alt="Main thread view of Mercado Libre's product detail pages after first round of optimizations.">
   <figcaption class="w-figcaption">

>       In the top waterfall of WPT there’s a long red bar (in the Page is Interactive row) between seconds 2 and 5. In the bottom waterfall, the bar has been broken into smaller pieces, occupying the main thread for shorter periods of time.

Reviewer: ```suggestion
      In the top waterfall of WPT there’s a long red bar (in the <b>Page is Interactive</b> row) between seconds 2 and 5. In the bottom waterfall, the bar has been broken into smaller pieces, occupying the main thread for shorter periods of time.
```

  </figcaption>     
</figure>

> Lighthouse showed a **50% reduction** in Max Potential First Input Delay:

Reviewer: ```suggestion
Lighthouse showed a **57% reduction** in Max Potential First Input Delay:
```
We should use the precise number. 1710ms to 750ms. I think that's 57% but you should double-check my math.

<figure class="w-figure">
  <img src="lighthouse-iteration-1.png" 
       alt="Lighthouse metrics in a PSI report for Mercado Libre's product detail pages after first round of optimizations.">
</figure>

## Second iteration

The team continued digging into long tasks in order to find subsequent improvements.

<figure class="w-figure">

>   <img src="main-thread-iteration-1-detail.png"

Reviewer: ```suggestion
  <img src="main-thread-iteration-1-detail.png" class="w-screenshot"
```

       alt="Detailed view of main thread view of Mercado Libre's product detail pages after first round of optimizations.">
   <figcaption class="w-figcaption">

>       The Waterfall (not pictured) helped Mercado Libre identify which libraries were using the main thread heavily (Browser Main Thread row) and the Page is Interactive row clearly shows that this main thread activity is blocking interactivity.

Reviewer: ```suggestion
      The Waterfall (not pictured) helped Mercado Libre identify which libraries were using the main thread heavily (<b>Browser Main Thread</b> row) and the <b>Page is Interactive</b> row clearly shows that this main thread activity is blocking interactivity.
```

  </figcaption>     
</figure>

Based on that information they decided to implement the following changes:

- Continue reducing the main bundle size to optimize compile and parse time (e.g. by removing duplicate dependencies throughout the different modules).
- Apply [code splitting](https://web.dev/reduce-javascript-payloads-with-code-splitting/) at component level, to divide JavaScript in smaller chunks and allow for smarter loading of the different components.
- Defer [component hydration](https://developers.google.com/web/updates/2019/02/rendering-on-the-web#rehydration) to allow for a smarter use of the main thread. This technique is commonly referred to as [partial hydration](https://addyosmani.com/blog/rehydration/).

## Measure impact

> The resulting Webpagetest trace showed even smaller chunks of JS execution:

Reviewer: ```suggestion
The resulting WebPageTest trace showed even smaller chunks of JS execution:
```

<figure class="w-figure">

>   <img src="main-thread-iteration-2.png"

Reviewer: ```suggestion
  <img src="main-thread-iteration-2.png" class="w-screenshot"
```

       alt="Main thread view of Mercado Libre's product detail pages after secoond round of optimizations.">
</figure>

And their Max Potential FID time in Lighthouse was reduced **by an additional 60%**:

<figure class="w-figure">

>   <img src="lighthouse-iteration-2.png"

Reviewer: ```suggestion
  <img src="lighthouse-iteration-2.png" class="w-screenshot"
```

---
layout: post
title: How Mercado Libre optimized for Web Vitals (TBT/FID)
subhead: Optimizing interactivity of product details pages for a 90% reduction in Max Potential FID in Lighthouse and a 9% improvement in FID in Chrome User Experience Report.
authors:
  - demianrenzulli
  - aranhacarlos
  - joanbaca

> date: 2020-09-15

Reviewer: ```suggestion
date: 2020-09-21
```

---
layout: post
title: How Mercado Libre optimized for Web Vitals (TBT/FID)
subhead: Optimizing interactivity of Mercado Libre's product details pages.
authors:
  - demianrenzulli
  - aranhacarlos
  - joanbaca

> date: 2020-09-01

Reviewer: ```suggestion
date: 2020-09-14
```

FID (First Input Delay) measures the time from when a user first interacts with a page (e.g. when
they click on a link) to the time when the browser is actually able to begin processing event
handlers in response to that interaction. A site that executes expensive Javascript code will likely
have several long tasks, which will end up negatively impacting FID.

To provide a good user experience, sites should strive to have a First Input Delay of less than 100
milliseconds:

<picture>
  <source srcset="../vitals/fid_8x2.svg" media="(min-width: 640px)">
  <img class="w-screenshot w-screenshot--filled"
      src="../vitals/fid_4x3.svg"
      alt="Good fid values are 2.5 seconds, poor values are greater than 4.0
            seconds and anything in between needs improvement">
</picture>

While Mercado Libre's site was performing well in most sections, they found in the [Chrome User
Experience Report](https://developers.google.com/web/tools/chrome-user-experience-report) that
product detail pages had a poor FID. Based on that information, they decided to focus their efforts
on improving the interactivity for product pages in the site.

<div class="w-columns">
<figure class="w-figure">

>   <img src="meli-pdp-mobile.png" 

Reviewer: ```suggestion
  <img src="meli-pdp-mobile.png" class="w-screenshot"
```

       alt="Mobile version of a Mercado Libre product detail page." width="141">
   <figcaption class="w-figcaption">
      Mobile version of a Mercado Libre product detail page.
  </figcaption>     
</figure>
<figure class="w-figure">

>   <img src="meli-pdp-desktop.png" 

Reviewer: ```suggestion
  <img src="meli-pdp-desktop.png" class="w-screenshot"
```

---
layout: post
title: How Mercado Libre optimized for Web Vitals (TBT/FID)

> subhead: Optimizing interactivity of Mercado Libre's product details pages.

Reviewer: ```suggestion
subhead: Optimizing interactivity of product details pages for a 90% reduction in First Input Delay and a 9% improvement in Chrome User Experience Report.
```
WDYT about mentioning the metric improvement here in the subtitle as well?

Author: It sounds like a good idea, but it should be: "for a 90% reduction in Reduction in Max Potential FID in Lighthouse and a 9% improvement in FID in Chrome User Experience Report."

I don't think I can apply a change to your suggestion from the UI and commit. Let me know if I should commit it on my end.

Reviewer: Fixed!
